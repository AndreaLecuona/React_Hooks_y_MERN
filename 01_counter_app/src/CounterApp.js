import React from 'react';
import PropTypes from 'prop-types';
import { useState } from 'react';

const CounterApp = ( { value = 0 } ) => {

    const [ counter, setCounter ] = useState( value );

    const handleAdd = () => {
        setCounter( counter + 1);
        // setCounter( (c) => c + 1); opciÃ³n2
    };

    const handleReset = () => {
        setCounter( value );
    };

    const handleSubstract = () => {
        setCounter( counter - 1 );
    };

    return (
        <div className="center">
        <h1>Counter App</h1>
        <h2> { counter } </h2>
        <button onClick={ handleSubstract }>- 1</button>
        <button onClick={ handleReset }>Reset</button>
        <button onClick={ handleAdd }>+ 1</button>
        </div>
    );
};

CounterApp.proptypes = {
    value: PropTypes.number
};

export default CounterApp;