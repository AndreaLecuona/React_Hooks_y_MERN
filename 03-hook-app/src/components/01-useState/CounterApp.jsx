import React, { useState } from 'react'
import './counter.css';

export const CounterApp = () => {

    const [state, setState] = useState({
        counter1: 10,
        counter2: 20,
        counter3: 30,
        counter4: 40
    });

    const { counter1, counter2 } = state;

    return (
        <div>
            <h1>Counter1 {counter1} </h1>
            <h1>Counter2 {counter2} </h1>
            <hr />

            {/* modificar un valor de un state sin modificar el resto */}
            <button className="btn btn-primary" onClick={() => setState( {
                ...state, //copia el estado anterior al nuevo estado
                counter1: counter1 + 1, //sobreescribe el valor que quiero
            } ) }>
                 +1 
            </button>
        </div >
    )
}
