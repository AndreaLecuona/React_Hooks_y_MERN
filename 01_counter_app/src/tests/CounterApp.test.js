import React from 'react';
import '@testing-library/jest-dom';
import { shallow } from "enzyme";

import CounterApp from '../CounterApp';

describe('Pruebas en CounterApp', () => {

    let wrapper = shallow( <CounterApp /> ); 
    //puedo subirlo a un scope de nivel mas alto para no repetirlo en cada prueba
    //debería estar solo incializado "let wrapper;" pero de esta forma inicia como undefined y más abajo en tests cuando quiero aplicarle métodos como find() o toMatchSnapshot() nunca aparecen sugeridos y hay que acordarse de memoria los nombres
    //definirlo asi como está y luego igual dentro del beforeEach es opcional, no es buena práctica repetir, pero en este caso es útil

    beforeEach( () => {

        wrapper = shallow( <CounterApp /> ); //lo reinicio antes de cada prueba, de lo contrario (como se ejecutan sincronicamente) las pruebas tomarían como valor inicial el resultado de la prueba anterior, y eso nos daría siempre fallos

    })

    test('Debe mostrar el componenente <CounterApp/> correctamente ', () => {

        expect( wrapper ).toMatchSnapshot();
    
    });

    test('<CounterApp/> debe mostrar el valor por defecto 0', () => {

        const wrapper = shallow( <CounterApp value={ 0 }/> ); //aca debo escribirlo porque le mando un valor especifico al props

        const valorContador = wrapper.find('h2').text().trim();

        expect( valorContador ).toBe( '0' );
    });
    
    test('Debe incrementar en una unidad con el boton +1 ', () => {
       
        wrapper.find('button').at(2).simulate('click'); //efectúo una simulacion de click

        const valorContador = wrapper.find('h2').text().trim(); //identifico el valor del h2 que ahora VA A ESTAR MODIFICADO por el evento simulado anterior (ejecucion sincronica)

        expect( valorContador ).toBe( '1' ); //comparo
    });

    test('Debe decrementar en una unidad con el boton -1 ', () => {
       
        wrapper.find('button').at(0).simulate('click');

        const valorContador = wrapper.find('h2').text().trim();

        expect( valorContador ).toBe( '-1' );
    });

    test('Debe volver el contador al valor por defecto con el boton Reset', () => {
       
        const wrapper = shallow( <CounterApp value={ 100 }/> ); //pedimos un valor original

        wrapper.find('button').at(2).simulate('click'); //simulamos el click +1 y cambiamos ese valor original a 101
        wrapper.find('button').at(2).simulate('click'); //simulamos el click +1 y cambiamos ese valor original a 102
        
        wrapper.find('button').at(1).simulate('click'); //simulamos click del reset

        const valorContador = wrapper.find('h2').text().trim(); //identifico valor de h2

        expect( valorContador ).toBe( '100' ); //comparo
    });
    

});
