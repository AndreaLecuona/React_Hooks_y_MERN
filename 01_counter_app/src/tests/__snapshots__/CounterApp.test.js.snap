// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en CounterApp Debe mostrar el componenente <CounterApp/> correctamente  1`] = `
<div
  className="center"
>
  <h1>
    Counter App
  </h1>
  <h2>
     
    0
     
  </h2>
  <button
    onClick={[Function]}
  >
    - 1
  </button>
  <button
    onClick={[Function]}
  >
    Reset
  </button>
  <button
    onClick={[Function]}
  >
    + 1
  </button>
</div>
`;
